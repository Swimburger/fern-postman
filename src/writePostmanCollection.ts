import { validateSchema } from "@fern-api/config-management-commons";
import { ExitStatusUpdate, GeneratorUpdate, LogLevel } from "@fern-fern/generator-exec-sdk/resources";
import * as GeneratorExecParsing from "@fern-fern/generator-exec-sdk/serialization";
import { IntermediateRepresentation } from "@fern-fern/ir-model/ir";
import { FernPostmanClient } from "@fern-fern/postman-sdk";
import * as PostmanParsing from "@fern-fern/postman-sdk/serialization";
import { readFile, writeFile } from "fs/promises";
import path from "path";
import { Collection, CollectionDefinition } from "postman-collection";
import { PostmanGeneratorConfigSchema } from "./config/schemas/PostmanGeneratorConfigSchema";
import { PublishConfigSchema } from "./config/schemas/PublishConfigSchema";
import { convertToPostmanCollection } from "./convertToPostmanCollection";
import { GeneratorNotificationService } from "./GeneratorNotificationService";
import { writePostmanGithubWorkflows } from "./writePostmanGithubWorkflows";

export const COLLECTION_OUTPUT_FILENAME = "collection.json";

export async function writePostmanCollection(pathToConfig: string): Promise<void> {
    try {
        console.log("Starting generator...");
        const configStr = await readFile(pathToConfig);
        console.log(`Read ${pathToConfig}`);
        const rawConfig = JSON.parse(configStr.toString());
        const config = GeneratorExecParsing.GeneratorConfig.parse(rawConfig);

        const postmanGeneratorConfig = await validateSchema(PostmanGeneratorConfigSchema, config.customConfig);
        console.log("Validated custom config");

        const generatorLoggingClient = new GeneratorNotificationService(config);
        console.log("Initialized generator logging client");

        try {
            await generatorLoggingClient.sendUpdate(
                GeneratorUpdate.init({
                    packagesToPublish: [],
                })
            );

            const ir = await loadIntermediateRepresentation(config.irFilepath);
            console.log(`Loaded intermediate representation from ${config.irFilepath}`);

            await generatorLoggingClient.sendUpdate(
                GeneratorUpdate.log({
                    level: LogLevel.Debug,
                    message: "Generating collection.json",
                })
            );
            const _collectionDefinition = convertToPostmanCollection(ir);
            const rawCollectionDefinition = PostmanParsing.PostmanCollectionSchema.json(_collectionDefinition);
            console.log("Converted ir to postman collection");

            await writeFile(
                path.join(config.output.path, COLLECTION_OUTPUT_FILENAME),
                JSON.stringify(rawCollectionDefinition, undefined, 2)
            );
            console.log(`Wrote postman collection to ${COLLECTION_OUTPUT_FILENAME}`);
            await generatorLoggingClient.sendUpdate(
                GeneratorUpdate.log({
                    level: LogLevel.Debug,
                    message: "Generated collection.json",
                })
            );

            const outputMode = config.output.mode;
            if (outputMode.type === "publish" && outputMode.publishTarget != null) {
                if (outputMode.publishTarget.type !== "postman") {
                    console.log(`Received incorrect publish config (type is ${outputMode.type}`);
                    await generatorLoggingClient.sendUpdate(
                        GeneratorUpdate.log({
                            level: LogLevel.Error,
                            message: `Received incorrect publish config (type is ${outputMode.type})`,
                        })
                    );
                    throw new Error("Received incorrect publish config!");
                }
                await publishCollection({
                    publishConfig: {
                        apiKey: outputMode.publishTarget.apiKey,
                        workspaceId: outputMode.publishTarget.workspaceId,
                    },
                    collectionDefinition: {
                        ...rawCollectionDefinition,
                        auth: rawCollectionDefinition.auth ?? undefined,
                    },
                });
            } else if (outputMode.type === "github") {
                console.log("Writing Github workflows...");
                await writePostmanGithubWorkflows({
                    config,
                    githubOutputMode: outputMode,
                });
            } else if (postmanGeneratorConfig?.publishing != null) {
                console.log("Publishing postman collection via legacy custom config...");
                await publishCollection({
                    publishConfig: postmanGeneratorConfig.publishing,
                    collectionDefinition: {
                        ...rawCollectionDefinition,
                        auth: rawCollectionDefinition.auth ?? undefined,
                    },
                });
            } else {
                console.log("Did not publish to postman or github");
            }

            await generatorLoggingClient.sendUpdate(GeneratorUpdate.exitStatusUpdate(ExitStatusUpdate.successful()));
        } catch (e) {
            console.log("Encountered error", e);
            await generatorLoggingClient.sendUpdate(
                GeneratorUpdate.exitStatusUpdate(
                    ExitStatusUpdate.error({
                        message: e instanceof Error ? e.message : "Encountered error",
                    })
                )
            );
        }
    } catch (e) {
        console.log("Encountered error", e);
        throw e;
    }
}

async function publishCollection({
    publishConfig,
    collectionDefinition,
}: {
    publishConfig: PublishConfigSchema;
    collectionDefinition: CollectionDefinition;
}) {
    console.log("Publishing postman collection...");
    const postman = new FernPostmanClient({
        auth: {
            apiKey: publishConfig.apiKey,
        },
    });
    const workspace = publishConfig.workspaceId != null ? publishConfig.workspaceId : undefined;
    console.log(`Workspace id is ${workspace}`);
    const getCollectionMetadataResponse = await postman.collection.getAllCollectionMetadata({
        workspace,
    });
    if (!getCollectionMetadataResponse.ok) {
        return;
    }
    const collectionsToUpdate = getCollectionMetadataResponse.body.collections.filter((collectionMetadata) => {
        return collectionMetadata.name === collectionDefinition.info?.name;
    });
    if (collectionsToUpdate.length === 0) {
        console.log("Creating new postman collection!");
        await postman.collection.createCollection({
            workspace,
            _body: { collection: new Collection(collectionDefinition) },
        });
    } else {
        collectionsToUpdate.forEach(async (collectionMetadata) => {
            console.log("Updating postman collection!");
            await postman.collection.updateCollection({
                collectionId: collectionMetadata.uid,
                _body: { collection: new Collection(collectionDefinition) },
            });
        });
    }
}

async function loadIntermediateRepresentation(pathToFile: string): Promise<IntermediateRepresentation> {
    return JSON.parse((await readFile(pathToFile)).toString());
}
