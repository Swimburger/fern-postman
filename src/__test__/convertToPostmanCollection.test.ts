/* eslint-disable jest/valid-title */
import {
    GeneratorConfig,
    GeneratorEnvironment,
    GithubPublishInfo,
    OutputMode,
} from "@fern-fern/generator-exec-sdk/resources";
import { execa } from "execa";
import { mkdir, readFile, rm, symlink, writeFile } from "fs/promises";
import path from "path";
import tmp from "tmp-promise";
import { COLLECTION_OUTPUT_FILENAME, writePostmanCollection } from "../writePostmanCollection";

const FIXTURES = ["test-api", "any-auth"];

describe("convertToPostman", () => {
    for (const fixture of FIXTURES) {
        // eslint-disable-next-line jest/valid-title
        const fixtureDir = path.join(__dirname, "fixtures");
        it(fixture, async () => {
            const tmpDir = await tmp.dir();

            const collectionPath = path.join(tmpDir.path, COLLECTION_OUTPUT_FILENAME);
            const configPath = path.join(tmpDir.path, "config.json");
            const irPath = path.join(tmpDir.path, "ir.json");

            // add symlink for easy access in VSCode
            const generatedDir = path.join(fixtureDir, "fern", fixture, "generated");
            await rm(generatedDir, { force: true, recursive: true });
            await mkdir(generatedDir);
            await symlink(collectionPath, path.join(generatedDir, COLLECTION_OUTPUT_FILENAME));

            const generatorConfig: GeneratorConfig = {
                dryRun: true,
                irFilepath: irPath,
                output: {
                    path: tmpDir.path,
                    mode: OutputMode.github({
                        repoUrl: "fern-api/fake",
                        version: "0.0.0",
                        publishInfo: GithubPublishInfo.postman({
                            apiKeyEnvironmentVariable: "API_KEY",
                            workspaceIdEnvironmentVariable: "WORKSPACE_ID",
                        }),
                    }),
                },
                publish: undefined,
                workspaceName: "fern",
                organization: "fern",
                customConfig: undefined,
                environment: GeneratorEnvironment.local(),
            };

            await writeFile(configPath, JSON.stringify(generatorConfig, undefined, 2));

            await execa("fern", ["ir", irPath, "--api", fixture], {
                cwd: fixtureDir,
            });

            await writePostmanCollection(configPath);

            const postmanCollection = (await readFile(collectionPath)).toString();

            expect(postmanCollection).toMatchSnapshot();
        });
    }
});
