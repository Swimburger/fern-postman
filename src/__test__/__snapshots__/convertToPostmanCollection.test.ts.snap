// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToPostman any-auth 1`] = `
"{
  \\"info\\": {
    \\"name\\": \\"Any Auth\\",
    \\"schema\\": \\"https://schema.getpostman.com/json/collection/v2.1.0/collection.json\\"
  },
  \\"variable\\": [
    {
      \\"key\\": \\"baseUrl\\",
      \\"value\\": \\"\\",
      \\"type\\": \\"string\\"
    },
    {
      \\"key\\": \\"username\\",
      \\"value\\": \\"\\",
      \\"type\\": \\"string\\"
    },
    {
      \\"key\\": \\"password\\",
      \\"value\\": \\"\\",
      \\"type\\": \\"string\\"
    },
    {
      \\"key\\": \\"headerAuth\\",
      \\"value\\": \\"\\",
      \\"type\\": \\"string\\"
    }
  ],
  \\"auth\\": {
    \\"type\\": \\"basic\\",
    \\"basic\\": [
      {
        \\"key\\": \\"username\\",
        \\"value\\": \\"{{username}}\\",
        \\"type\\": \\"string\\"
      },
      {
        \\"key\\": \\"password\\",
        \\"value\\": \\"{{password}}\\",
        \\"type\\": \\"string\\"
      }
    ]
  },
  \\"item\\": [
    {
      \\"_type\\": \\"container\\",
      \\"description\\": null,
      \\"name\\": \\"Blog\\",
      \\"item\\": [
        {
          \\"_type\\": \\"endpoint\\",
          \\"name\\": \\"Create Post\\",
          \\"request\\": {
            \\"description\\": null,
            \\"url\\": {
              \\"raw\\": \\"{{baseUrl}}/create\\",
              \\"host\\": [
                \\"{{baseUrl}}\\"
              ],
              \\"path\\": [
                \\"create\\"
              ],
              \\"query\\": [],
              \\"variable\\": []
            },
            \\"header\\": [
              {
                \\"key\\": \\"My-Header\\",
                \\"value\\": \\"{{headerAuth}}\\",
                \\"type\\": \\"string\\",
                \\"description\\": null
              }
            ],
            \\"method\\": \\"POST\\",
            \\"auth\\": null,
            \\"body\\": {
              \\"mode\\": \\"raw\\",
              \\"raw\\": \\"{\\\\n    \\\\\\"title\\\\\\": \\\\\\"example\\\\\\"\\\\n}\\",
              \\"options\\": {
                \\"raw\\": {
                  \\"language\\": \\"json\\"
                }
              }
            }
          },
          \\"response\\": []
        },
        {
          \\"_type\\": \\"endpoint\\",
          \\"name\\": \\"Get Post\\",
          \\"request\\": {
            \\"description\\": null,
            \\"url\\": {
              \\"raw\\": \\"{{baseUrl}}/:postId\\",
              \\"host\\": [
                \\"{{baseUrl}}\\"
              ],
              \\"path\\": [
                \\":postId\\"
              ],
              \\"query\\": [],
              \\"variable\\": [
                {
                  \\"key\\": \\"postId\\",
                  \\"value\\": \\"\\",
                  \\"description\\": null
                }
              ]
            },
            \\"header\\": [
              {
                \\"key\\": \\"My-Header\\",
                \\"value\\": \\"{{headerAuth}}\\",
                \\"type\\": \\"string\\",
                \\"description\\": null
              }
            ],
            \\"method\\": \\"GET\\",
            \\"auth\\": null,
            \\"body\\": null
          },
          \\"response\\": []
        }
      ]
    }
  ]
}"
`;

exports[`convertToPostman loop-test-api 1`] = `
"{
  \\"info\\": {
    \\"name\\": \\"Loop API\\",
    \\"schema\\": \\"https://schema.getpostman.com/json/collection/v2.1.0/collection.json\\"
  },
  \\"variable\\": [
    {
      \\"key\\": \\"baseUrl\\",
      \\"value\\": \\"prod.fern.com\\",
      \\"type\\": \\"string\\"
    },
    {
      \\"key\\": \\"apiKey\\",
      \\"value\\": \\"\\",
      \\"type\\": \\"string\\"
    }
  ],
  \\"auth\\": {
    \\"type\\": \\"apikey\\",
    \\"apikey\\": [
      {
        \\"key\\": \\"value\\",
        \\"value\\": \\"{{apiKey}}\\",
        \\"type\\": \\"string\\"
      },
      {
        \\"key\\": \\"key\\",
        \\"value\\": \\"X-API-Key\\",
        \\"type\\": \\"string\\"
      },
      {
        \\"key\\": \\"in\\",
        \\"value\\": \\"header\\",
        \\"type\\": \\"string\\"
      }
    ]
  },
  \\"item\\": [
    {
      \\"_type\\": \\"container\\",
      \\"description\\": null,
      \\"name\\": \\"Orders\\",
      \\"item\\": [
        {
          \\"_type\\": \\"container\\",
          \\"description\\": null,
          \\"name\\": \\"Pet\\",
          \\"item\\": [
            {
              \\"_type\\": \\"endpoint\\",
              \\"name\\": \\"Create Pet\\",
              \\"request\\": {
                \\"description\\": \\"Creates a pet and returns pet id.\\",
                \\"url\\": {
                  \\"raw\\": \\"{{baseUrl}}/create\\",
                  \\"host\\": [
                    \\"{{baseUrl}}\\"
                  ],
                  \\"path\\": [
                    \\"create\\"
                  ],
                  \\"query\\": [],
                  \\"variable\\": []
                },
                \\"header\\": [
                  {
                    \\"key\\": \\"X-API-Key\\",
                    \\"value\\": \\"{{apiKey}}\\",
                    \\"type\\": \\"string\\",
                    \\"description\\": null
                  },
                  {
                    \\"key\\": \\"x-service-header\\",
                    \\"description\\": null,
                    \\"type\\": \\"text\\",
                    \\"value\\": \\"\\\\\\"example\\\\\\"\\"
                  },
                  {
                    \\"key\\": \\"x-endpoint-header\\",
                    \\"description\\": \\"I'm a header!\\",
                    \\"type\\": \\"text\\",
                    \\"value\\": \\"\\\\\\"example\\\\\\"\\"
                  }
                ],
                \\"method\\": \\"POST\\",
                \\"auth\\": null,
                \\"body\\": {
                  \\"mode\\": \\"raw\\",
                  \\"raw\\": \\"{\\\\n    \\\\\\"type\\\\\\": {\\\\n        \\\\\\"type\\\\\\": \\\\\\"CAT\\\\\\",\\\\n        \\\\\\"friends\\\\\\": [\\\\n            null\\\\n        ]\\\\n    },\\\\n    \\\\\\"name\\\\\\": \\\\\\"example\\\\\\"\\\\n}\\",
                  \\"options\\": {
                    \\"raw\\": {
                      \\"language\\": \\"json\\"
                    }
                  }
                }
              },
              \\"response\\": []
            }
          ]
        }
      ]
    }
  ]
}"
`;

exports[`convertToPostman test-api 1`] = `
"{
  \\"info\\": {
    \\"name\\": \\"Blog Post API\\",
    \\"schema\\": \\"https://schema.getpostman.com/json/collection/v2.1.0/collection.json\\"
  },
  \\"variable\\": [
    {
      \\"key\\": \\"baseUrl\\",
      \\"value\\": \\"prod.fern.com\\",
      \\"type\\": \\"string\\"
    },
    {
      \\"key\\": \\"apiKey\\",
      \\"value\\": \\"\\",
      \\"type\\": \\"string\\"
    }
  ],
  \\"auth\\": {
    \\"type\\": \\"apikey\\",
    \\"apikey\\": [
      {
        \\"key\\": \\"value\\",
        \\"value\\": \\"{{apiKey}}\\",
        \\"type\\": \\"string\\"
      },
      {
        \\"key\\": \\"key\\",
        \\"value\\": \\"X-API-Key\\",
        \\"type\\": \\"string\\"
      },
      {
        \\"key\\": \\"in\\",
        \\"value\\": \\"header\\",
        \\"type\\": \\"string\\"
      }
    ]
  },
  \\"item\\": [
    {
      \\"_type\\": \\"container\\",
      \\"description\\": null,
      \\"name\\": \\"Blog\\",
      \\"item\\": [
        {
          \\"_type\\": \\"endpoint\\",
          \\"name\\": \\"Create Blog Post\\",
          \\"request\\": {
            \\"description\\": \\"Creates a blog post and returns blog post id.\\",
            \\"url\\": {
              \\"raw\\": \\"{{baseUrl}}/create\\",
              \\"host\\": [
                \\"{{baseUrl}}\\"
              ],
              \\"path\\": [
                \\"create\\"
              ],
              \\"query\\": [],
              \\"variable\\": []
            },
            \\"header\\": [
              {
                \\"key\\": \\"X-API-Key\\",
                \\"value\\": \\"{{apiKey}}\\",
                \\"type\\": \\"string\\",
                \\"description\\": null
              },
              {
                \\"key\\": \\"x-service-header\\",
                \\"description\\": null,
                \\"type\\": \\"text\\",
                \\"value\\": \\"\\\\\\"example\\\\\\"\\"
              },
              {
                \\"key\\": \\"x-endpoint-header\\",
                \\"description\\": \\"I'm a header!\\",
                \\"type\\": \\"text\\",
                \\"value\\": \\"\\\\\\"example\\\\\\"\\"
              }
            ],
            \\"method\\": \\"POST\\",
            \\"auth\\": null,
            \\"body\\": {
              \\"mode\\": \\"raw\\",
              \\"raw\\": \\"{\\\\n    \\\\\\"title\\\\\\": \\\\\\"example\\\\\\",\\\\n    \\\\\\"author\\\\\\": {\\\\n        \\\\\\"type\\\\\\": \\\\\\"name\\\\\\",\\\\n        \\\\\\"value\\\\\\": \\\\\\"George Washington\\\\\\"\\\\n    },\\\\n    \\\\\\"postType\\\\\\": \\\\\\"LONG\\\\\\"\\\\n}\\",
              \\"options\\": {
                \\"raw\\": {
                  \\"language\\": \\"json\\"
                }
              }
            }
          },
          \\"response\\": []
        },
        {
          \\"_type\\": \\"endpoint\\",
          \\"name\\": \\"Get Post\\",
          \\"request\\": {
            \\"description\\": \\"Gets a blog post.\\",
            \\"url\\": {
              \\"raw\\": \\"{{baseUrl}}/:postId\\",
              \\"host\\": [
                \\"{{baseUrl}}\\"
              ],
              \\"path\\": [
                \\":postId\\"
              ],
              \\"query\\": [],
              \\"variable\\": [
                {
                  \\"key\\": \\"postId\\",
                  \\"description\\": null,
                  \\"value\\": \\"123e4567-e89b-12d3-a456-426614174000\\"
                }
              ]
            },
            \\"header\\": [
              {
                \\"key\\": \\"X-API-Key\\",
                \\"value\\": \\"{{apiKey}}\\",
                \\"type\\": \\"string\\",
                \\"description\\": null
              },
              {
                \\"key\\": \\"x-service-header\\",
                \\"description\\": null,
                \\"type\\": \\"text\\",
                \\"value\\": \\"\\\\\\"I'm a service header!\\\\\\"\\"
              },
              {
                \\"type\\": \\"text\\",
                \\"key\\": \\"Content-Type\\",
                \\"value\\": \\"application/json\\"
              }
            ],
            \\"method\\": \\"GET\\",
            \\"auth\\": null,
            \\"body\\": null
          },
          \\"response\\": [
            {
              \\"name\\": \\"Success\\",
              \\"status\\": \\"OK\\",
              \\"code\\": 200,
              \\"originalRequest\\": {
                \\"description\\": \\"Gets a blog post.\\",
                \\"url\\": {
                  \\"raw\\": \\"{{baseUrl}}/:postId\\",
                  \\"host\\": [
                    \\"{{baseUrl}}\\"
                  ],
                  \\"path\\": [
                    \\":postId\\"
                  ],
                  \\"query\\": [],
                  \\"variable\\": [
                    {
                      \\"key\\": \\"postId\\",
                      \\"description\\": null,
                      \\"value\\": \\"123e4567-e89b-12d3-a456-426614174000\\"
                    }
                  ]
                },
                \\"header\\": [
                  {
                    \\"key\\": \\"X-API-Key\\",
                    \\"value\\": \\"{{apiKey}}\\",
                    \\"type\\": \\"string\\",
                    \\"description\\": null
                  },
                  {
                    \\"key\\": \\"x-service-header\\",
                    \\"description\\": null,
                    \\"type\\": \\"text\\",
                    \\"value\\": \\"\\\\\\"I'm a service header!\\\\\\"\\"
                  },
                  {
                    \\"type\\": \\"text\\",
                    \\"key\\": \\"Content-Type\\",
                    \\"value\\": \\"application/json\\"
                  }
                ],
                \\"method\\": \\"GET\\",
                \\"auth\\": null,
                \\"body\\": null
              },
              \\"description\\": null,
              \\"body\\": \\"{\\\\n    \\\\\\"writtenDate\\\\\\": \\\\\\"2022-12-15T22:06:28+00:00\\\\\\",\\\\n    \\\\\\"id\\\\\\": \\\\\\"123e4567-e89b-12d3-a456-426614174000\\\\\\",\\\\n    \\\\\\"type\\\\\\": \\\\\\"LONG\\\\\\",\\\\n    \\\\\\"title\\\\\\": \\\\\\"My Post\\\\\\",\\\\n    \\\\\\"author\\\\\\": {\\\\n        \\\\\\"type\\\\\\": \\\\\\"anonymous\\\\\\"\\\\n    },\\\\n    \\\\\\"content\\\\\\": \\\\\\"I'm a blog post\\\\\\"\\\\n}\\",
              \\"_postman_previewlanguage\\": \\"json\\"
            },
            {
              \\"name\\": \\"Post Not Found Error\\",
              \\"status\\": \\"Post Not Found Error\\",
              \\"code\\": 400,
              \\"originalRequest\\": {
                \\"description\\": \\"Gets a blog post.\\",
                \\"url\\": {
                  \\"raw\\": \\"{{baseUrl}}/:postId\\",
                  \\"host\\": [
                    \\"{{baseUrl}}\\"
                  ],
                  \\"path\\": [
                    \\":postId\\"
                  ],
                  \\"query\\": [],
                  \\"variable\\": [
                    {
                      \\"key\\": \\"postId\\",
                      \\"description\\": null,
                      \\"value\\": \\"123e4567-e89b-12d3-a456-426614174000\\"
                    }
                  ]
                },
                \\"header\\": [
                  {
                    \\"key\\": \\"X-API-Key\\",
                    \\"value\\": \\"{{apiKey}}\\",
                    \\"type\\": \\"string\\",
                    \\"description\\": null
                  },
                  {
                    \\"key\\": \\"x-service-header\\",
                    \\"description\\": null,
                    \\"type\\": \\"text\\",
                    \\"value\\": \\"\\\\\\"I'm a service header!\\\\\\"\\"
                  },
                  {
                    \\"type\\": \\"text\\",
                    \\"key\\": \\"Content-Type\\",
                    \\"value\\": \\"application/json\\"
                  }
                ],
                \\"method\\": \\"GET\\",
                \\"auth\\": null,
                \\"body\\": null
              },
              \\"description\\": null,
              \\"body\\": \\"{\\\\n    \\\\\\"id\\\\\\": \\\\\\"123e4567-e89b-12d3-a456-426614174000\\\\\\"\\\\n}\\",
              \\"_postman_previewlanguage\\": \\"json\\"
            }
          ]
        }
      ]
    },
    {
      \\"_type\\": \\"container\\",
      \\"description\\": \\"I'm some docs about MyFolder\\",
      \\"name\\": \\"My Folder\\",
      \\"item\\": [
        {
          \\"_type\\": \\"container\\",
          \\"description\\": null,
          \\"name\\": \\"My Folder\\",
          \\"item\\": [
            {
              \\"_type\\": \\"container\\",
              \\"description\\": null,
              \\"name\\": \\"Nested\\",
              \\"item\\": [
                {
                  \\"_type\\": \\"endpoint\\",
                  \\"name\\": \\"Get\\",
                  \\"request\\": {
                    \\"description\\": null,
                    \\"url\\": {
                      \\"raw\\": \\"{{baseUrl}}\\",
                      \\"host\\": [
                        \\"{{baseUrl}}\\"
                      ],
                      \\"path\\": [],
                      \\"query\\": [],
                      \\"variable\\": []
                    },
                    \\"header\\": [
                      {
                        \\"key\\": \\"X-API-Key\\",
                        \\"value\\": \\"{{apiKey}}\\",
                        \\"type\\": \\"string\\",
                        \\"description\\": null
                      },
                      {
                        \\"key\\": \\"x-service-header\\",
                        \\"description\\": null,
                        \\"type\\": \\"text\\",
                        \\"value\\": \\"\\\\\\"example\\\\\\"\\"
                      }
                    ],
                    \\"method\\": \\"GET\\",
                    \\"auth\\": null,
                    \\"body\\": null
                  },
                  \\"response\\": []
                }
              ]
            }
          ]
        },
        {
          \\"_type\\": \\"container\\",
          \\"description\\": null,
          \\"name\\": \\"Nested\\",
          \\"item\\": [
            {
              \\"_type\\": \\"endpoint\\",
              \\"name\\": \\"Get\\",
              \\"request\\": {
                \\"description\\": null,
                \\"url\\": {
                  \\"raw\\": \\"{{baseUrl}}\\",
                  \\"host\\": [
                    \\"{{baseUrl}}\\"
                  ],
                  \\"path\\": [],
                  \\"query\\": [],
                  \\"variable\\": []
                },
                \\"header\\": [
                  {
                    \\"key\\": \\"X-API-Key\\",
                    \\"value\\": \\"{{apiKey}}\\",
                    \\"type\\": \\"string\\",
                    \\"description\\": null
                  },
                  {
                    \\"key\\": \\"x-service-header\\",
                    \\"description\\": null,
                    \\"type\\": \\"text\\",
                    \\"value\\": \\"\\\\\\"example\\\\\\"\\"
                  }
                ],
                \\"method\\": \\"GET\\",
                \\"auth\\": null,
                \\"body\\": null
              },
              \\"response\\": []
            }
          ]
        }
      ]
    }
  ]
}"
`;
